catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/suggestion/MgrsCoordinateProcessor.java

```diff
@@ -0,0 +1,254 @@
+/**
+ * Copyright (c) Codice Foundation
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.catalog.ui.query.suggestion;
+
+import static java.lang.String.format;
+import static org.codice.ddf.catalog.ui.query.suggestion.LatLon.from;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import java.text.ParseException;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import org.codice.ddf.spatial.geocoding.Suggestion;
+import org.codice.usng4j.BoundingBox;
+import org.codice.usng4j.CoordinateSystemTranslator;
+import org.codice.usng4j.UsngCoordinate;
+import org.codice.usng4j.impl.CoordinateSystemTranslatorImpl;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MgrsCoordinateProcessor {
+  private static final Logger LOGGER = LoggerFactory.getLogger(MgrsCoordinateProcessor.class);
+
+  private static final Pattern PATTERN_MGRS_ZONE = Pattern.compile("\\d\\d?[CDEFGHJKLMNPQRSTUVWX]");
+
+  private static final Pattern PATTERN_MGRS_100KM =
+      Pattern.compile("[ABCDEFGHJKLMNPQRSTUVWXYZ][ABCDEFGHJKLMNPQRSTUV]");
+
+  private static final Pattern PATTERN_MGRS_NUMERIC =
+      Pattern.compile("(\\d){1,5}[ ]*(\\d){1,5}[ ]*");
+
+  private static final Pattern PATTERN_MGRS_ANYPART =
+      Pattern.compile(
+          format(
+              "(%s)|(%s)|(%s)",
+              PATTERN_MGRS_ZONE.pattern(),
+              PATTERN_MGRS_100KM.pattern(),
+              PATTERN_MGRS_NUMERIC.pattern()));
+
+  private static final Integer MIN_EASTING_NORTHING_LENGTH = 6;
+
+  private static final Integer MIN_EASTING_LENGTH = 5;
+
+  private enum MgrsPartType {
+    NONE,
+    ZONE,
+    HUNDRED_KM,
+    NUMERIC
+  }
+
+  private final CoordinateSystemTranslator translator = new CoordinateSystemTranslatorImpl();
+
+  public List<Suggestion> enhanceResults(List<Suggestion> results, String query) {
+    LinkedList<Suggestion> enhanced = new LinkedList<>(results);
```

Comment:
should use list on the left side; why are we using a linked list?

Reply:
For suggestions we want to add to the beginning of the list but that would entail a complete shifting of all the elements for every suggestion added. Using a linked list speeds this up due to head of list insertion complexity `O(1)`. 

I kept the var as a `LinkedList` so I could leverage the `addFirst(literal)` method instead of `add(0, literal)`. Since the return type is still `List` I wasn't concerned.